### PR TITLE
[Diagnostics] Improve diagnostics involving macros

### PR DESCRIPTION
**Pull request by bnbarham**
The main change here is to update the `in expansion of macro` note location for attached macros to that of the attribute instead of the declaration. But as that then makes the declaration it's on less obvious, I've added its name to the message as well. Since we just changed the location of `macro_attached_to_invalid_decl` I also added the declaration name there as well.

### CODE REVIEW

**Comment by bnbarham** [comment]
@swift-ci please test

**Comment by bnbarham** [comment]
@swift-ci please test

File: lib/AST/DiagnosticEngine.cpp [hunk 1]
```diff
@@ -1406,11 +1406,21 @@ DiagnosticEngine::getGeneratedSourceBufferNotes(SourceLoc loc) {
...
       DeclName macroName = getGeneratedSourceInfoMacroName(*generatedInfo);
 
-      Diagnostic expansionNote(diag::in_macro_expansion, macroName);
-      expansionNote.setLoc(origRange.Start);
+      // If it was an expansion of an attached macro, increase the range to
+      // include the decl's attributes. Also add the name of the decl the macro
+      // is attached to.
+      CustomAttr *attachedAttr = generatedInfo->attachedMacroCustomAttr;
+      Decl *attachedDecl = attachedAttr ? expansionNode.dyn_cast<Decl *>() : nullptr;
+      SourceRange origRange = attachedDecl ? attachedDecl->getSourceRangeIncludingAttrs() : expansionNode.getSourceRange();
```
Comment (xedin) [inline lib/AST/DiagnosticEngine.cpp:1416]: Nit: This looks unformatted to me.

**Review by xedin — APPROVED** [review]

File: lib/AST/DiagnosticEngine.cpp (hunk 1 above)
Comment (bnbarham) [reply lib/AST/DiagnosticEngine.cpp]: Indeed, fixed.

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by hborla — APPROVED** [review]

**Pull request merged**